src/inmanta/execute/dataflow/root_cause.py

```diff
@@ -0,0 +1,122 @@
+"""
+    Copyright 2020 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from typing import FrozenSet, Iterable, List, Set
+
+from inmanta.execute.dataflow import AssignableNode, AttributeNode, AttributeNodeReference
+
+
+class UnsetRootCauseAnalyzer:
+    """
+        Analyzes the root causes for attributes being unset among a collection of attribute nodes.
+        The main entrypoint for this class is the root_causes method.
+    """
+
+    def __init__(self, nodes: Iterable[AttributeNode]) -> None:
+        self.nodes: FrozenSet[AttributeNode] = frozenset(nodes)
+
+    def root_causes(self) -> Set[AttributeNode]:
+        """
+            Returns the root causes from this instances' set of attribute nodes. An attribute node c
+            is defined as the cause for an other attribute node n iff c being unset leads to n
+            being unset.
+            Formally, the relation is_cause(c, x) is defined by three rules:
+                1. is_cause(c, x) <- c in `x = c` in graph
+                    (If `x = c` then c is responsible for x receiving a value)
+                2. is_cause(c, x) <- exists y: is_cause(y, x) and is_cause(c, y)
```

Comment:
I hesitated. I opted for (c, y) at the end because that's the one that will need to be refined next and this sort of tail recursion was consistent with the other rules. Knowing this, do you still prefer it the other way around?

Reply:
Yes, I like having the flow apparent, but is is a minor preference.